feat(`v1`): Migration guide (#922)
foundry-rs/book: src/migrating-to-v1/deprecated-features.md

## Deprecated features

> With foundry v1, we've started deprecating features that are considered anti-patterns or might be misleading and confusing for developers. Deprecated means that while you can continue using these features, it is now a discouraged practice and you should plan on migrating away from using these.

Reviewer: Foundry

Author: Done in https://github.com/foundry-rs/book/pull/922/commits/0622765622b80670ca2b838eff1b368440516b65

> ### Removing the Invariant keyword

Reviewer: `invariant` prefix?

Author: Done in https://github.com/foundry-rs/book/pull/922/commits/0622765622b80670ca2b838eff1b368440516b65

The invariant keyword has now been deprecated, and the new keyword is now `statefulFuzz`. This means that now writing tests in this manner is valid:

```solidity
/// Old, deprecated way of declaring an invariant test
function invariantTestEq() public {
    // Assert your invariants...
}

/// New
function statefulFuzzTestEq() public  {
    // Assert your invariants
}
```

> ### Deprecating testFail

Reviewer 2: ```suggestion
### `testFail`
```

Using `testFail` to write failing tests is now discouraged. While it's usage is extremely common, we've decided to deprecate it because it can introduce footguns in a test.

> The now-recommended pattern is to refactor the actions that will make your test revert and subsequently fail in an utility function, and use `expectRevert` to ensure that this function call failed. This way, the test failure is expected explicitly, instead of introducing the possibility that the test fails in an unintended way silently.

Reviewer 2: Similarly, I think this has always been recommended. And we can simplify this a bit since the "refactor" part may not be necessary and has already been covered in the expectRevert section
```suggestion
A better pattern is to use `expectRevert` to ensure that a function call reverted in a specific way. This way, the test failure is expected explicitly, removing the possibility that the test fails in an unintended, undetectable way.
```

An example:

```solidity

contract Mock {
    function revert_() public {
        revert("This reverts");
    }
}

contract TestFailDeprecated is Test {
    Mock public mock;

    function setUp() public {
        mock = new Mock();
    }

    /// Deprecated way.
    function testFailReverter() public {
        mock.revert_();
    }

    /// New way, without using testFail.
    /// The call to revert_ has been refactored and is now expected to fail
    /// with expectRevert()

>     function testReverterReverts() public {

Reviewer 2: We should also start making sure the test names in these code snippets match what's in the best practices page, for consistency
```suggestion
    function test_RevertIf_AlwaysReverts() public {
```

>         vm.expectRevert();

Reviewer 2: let's avoid using the empty args version in the examples, since (like `testFail`) this is weaker and should only be used when necessary

>     function test_RevertIf_AlwaysReverts() public {

Reviewer: `test_revertIf`..

Author: Done in https://github.com/foundry-rs/book/pull/922/commits/0622765622b80670ca2b838eff1b368440516b65

Reviewer 2: The best practices page of the book uses `test_RevertIf_AlwaysReverts` style though

>         vm.expectRevert("This reverts);

Reviewer: "This reverts"

Author: Done in https://github.com/foundry-rs/book/pull/922/commits/0622765622b80670ca2b838eff1b368440516b65

great catch!

        // Call using `this` to increase depth
        this.exposed_call_Reverter();
    }

    function exposed_callReverter() public {
        mock.revert_();
    }

}
```

### Removing cheatcodes support on some precompiles

> It is now impossible to use the following cheatcodes on precompiles: `etch`, `load`, `store`, `deal`.

Reviewer 2: let's add rationale as to why
